### PR TITLE
Feature set attribute gloidx array

### PR DESCRIPTION
**Pull request by holke**
**_Describe your changes here:_**

A new setter and getter for cmesh attributes of type t8_gloidx_t array.
That is multiple t8_gloidx_t entries.

- Implement t8_cmesh_set_attribute_gloidx_array 
- Implement t8_cmesh_get_attribute_gloidx_array 
- Implement a test for set/get with a cmesh with 2 trees.
- Also made minor changes on the existing attribute functions that i believe to be small enough to inlcude here, even though
they could be their own PR.

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_cmesh.h [hunk 1]
```diff
@@ -277,10 +277,37 @@ t8_cmesh_set_attribute (t8_cmesh_t cmesh, t8_gloidx_t gtree_id, int package_id,
...
+ *                              In both cases a copy of the data is used by t8_code after t8_cmesh_commit.
+ * \note You can also use \ref t8_cmesh_set_attribute, but we recommend using this
+ *       specialized function for arrays.
+ * \note If an attribute with the given package_id and key already exists, then it will get overwritten.
+ * \note We do not store the number of data entries \a data_count of the attribute array.
+ *       You can keep track of the data count yourself by using another attribute.
+ */
+void
+t8_cmesh_set_attribute_gloidx_array (t8_cmesh_t cmesh, t8_gloidx_t gtree_id, int package_id, int key,
+                                     const t8_gloidx_t *data, size_t data_count, int data_persists);
```
Comment (Davknapp) [inline src/t8_cmesh.h:309]:
```suggestion
                                     const t8_gloidx_t *data, const size_t data_count, int data_persists);
```

File: src/t8_cmesh.h [hunk 2]
```diff
@@ -632,14 +659,33 @@ t8_cmesh_get_tree_vertices (t8_cmesh_t cmesh, t8_locidx_t ltreeid);
...
+ * \return         The attribute pointer of the tree \a ltree_id or NULL if the attribute is not found.
+ * \note \a cmesh must be committed before calling this function.
+ * \note No check is performed whether the attribute actually stored \a data_count many entries since
+ *       we do not store the number of data entries of the attribute array.
+ *       You can keep track of the data count yourself by using another attribute.
+ * \see t8_cmesh_set_attribute_gloidx_array
+ */
+t8_gloidx_t *
+t8_cmesh_get_attribute_gloidx_array (t8_cmesh_t cmesh, int package_id, int key, t8_locidx_t ltree_id,
+                                     size_t data_count);
```
Comment (Davknapp) [inline src/t8_cmesh.h:687]:
```suggestion
                                     const size_t data_count);
```
It is possible to make ltree_id constant, too, but I propose to do this in another PR, since it requires changes outside of the topic of this PR

File: src/t8_cmesh/t8_cmesh.c [hunk 3]
```diff
@@ -384,18 +392,39 @@ t8_cmesh_get_tree_vertices (t8_cmesh_t cmesh, t8_locidx_t ltreeid)
...
+ *                              of the corresponding call to \ref t8_cmesh_set_attribute_gloidx_array
+ * \return         The attribute pointer of the tree \a ltree_id or NULL if the attribute is not found.
+ * \note \a cmesh must be committed before calling this function.
+ * \note No check is performed whether the attribute actually stored \a data_count many entries since
+ *       we do not store the number of data entries of the attribute array.
+ *       You can keep track of the data count yourself by using another attribute.
+ * \see t8_cmesh_set_attribute_gloidx_array
+ */
+t8_gloidx_t *
+t8_cmesh_get_attribute_gloidx_array (t8_cmesh_t cmesh, int package_id, int key, t8_locidx_t ltree_id, size_t data_count)
```
Comment (Davknapp) [inline src/t8_cmesh/t8_cmesh.c:422]:
```suggestion
t8_cmesh_get_attribute_gloidx_array (t8_cmesh_t cmesh, int package_id, int key, t8_locidx_t ltree_id, const size_t data_count)
```

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx [hunk 4]
```diff
@@ -0,0 +1,135 @@
+/*
+  This file is part of t8code.
+  t8code is a C library to manage a collection (a forest) of multiple
+  connected adaptive space-trees of general element classes in parallel.
+
+  Copyright (C) 2015 the developers
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:6]:
```suggestion
  Copyright (C) 2024 the developers
```

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+/** Check attribute values of cmeshes against reference values. */
+TEST_P (cmesh_attribute_gloidx_array, check_values_data_not_persist)
+{
+  const t8_locidx_t tree_id = 0;
+  get_entries = t8_cmesh_get_attribute_gloidx_array (cmesh, t8_get_package_id (), T8_CMESH_NEXT_POSSIBLE_KEY, tree_id,
+                                                     num_entries);
+
+  /* If we did not store any values, we except to get the NULL pointer back. */
+  if (entries == NULL) {
+    EXPECT_TRUE (get_entries == NULL);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:93]:
```suggestion
    EXPECT_EQ (get_entries, nullptr);
```

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    /* Number of entries must be < 0 in that case and we cannot continue the test otherwise. */
+    ASSERT_EQ (num_entries, 0);
+  }
+  else {
+    /* Otherwise it must not be NULL and we abort the test if it is. */
+    ASSERT_TRUE (get_entries != NULL);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:99]:
```suggestion
    ASSERT_NE (get_entries, nullptr);
```

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+/** Check attribute values of cmeshes against reference values. */
+TEST_P (cmesh_attribute_gloidx_array, check_values_data_persist)
+{
+  const t8_locidx_t tree_id = 1;
+  get_entries = t8_cmesh_get_attribute_gloidx_array (cmesh, t8_get_package_id (), T8_CMESH_NEXT_POSSIBLE_KEY, tree_id,
+                                                     num_entries);
+
+  /* If we did not store any values, we except to get the NULL pointer back. */
+  if (entries == NULL) {
+    EXPECT_TRUE (get_entries == NULL);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:117]:
```suggestion
    EXPECT_EQ (get_entries, nullptr);
```

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    /* Number of entries must be < 0 in that case and we cannot continue the test otherwise. */
+    ASSERT_EQ (num_entries, 0);
+  }
+  else {
+    /* Otherwise it must not be NULL and we abort the test if it is. */
+    ASSERT_TRUE (get_entries != NULL);
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:123]:
```suggestion
    ASSERT_NE (get_entries, nullptr);
```

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    /* Set the array as attribute twice. Once with data_persist and once without. */
+    /* Attribute at tree 0, data_persist = 0 */
+    t8_gloidx_t tree_with_attribute = 0;
+    int data_persists = 0;
+    t8_cmesh_set_attribute_gloidx_array (cmesh, tree_with_attribute, t8_get_package_id (), T8_CMESH_NEXT_POSSIBLE_KEY,
+                                         entries, num_entries, data_persists);
+
+    /* Attribute at tree 1, data_persist = 1 */
+    tree_with_attribute = 1;
+    data_persists = 1;
```
Comment (Davknapp) [inline test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:62]: you can use ```data_persists``` as an additional parameter for the test using ```Combine(generator0, generator1, ...)```. That way, we can simplify the tests down to a single test (the two tests are the same (or I missed a very small difference), which avoids code-duplication.

Comment (holke) [reply test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:63]: Please check what i did.
i combined the tests into 1.
I still think it good to have 2 trees in the cmesh to cover multiple trees.
Both trees store the attribute but with different data_persist.
we then run the test twice, once for each tree.

**Review by Davknapp — APPROVED** [review]

File: test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx (hunk 9 above)
Comment (Davknapp) [reply test/t8_cmesh/t8_gtest_attribute_gloidx_array.cxx:71]: Yes, this solution is also in the spirit of my remark.
